### PR TITLE
Bind switch expression with checked-aware binder

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/61843
Fixes https://github.com/dotnet/roslyn/issues/60419

The issue is that `Binder.BindSwitchExpression` is dropping the binder `Flags`, so we're losing the `OverflowChecks.Enabled` flag set by `Binder.BindCheckedExpression`.

```
        private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, BindingDiagnosticBag diagnostics)
        {
            RoslynDebug.Assert(node is { });
            Binder? switchBinder = this.GetBinder(node);
            RoslynDebug.Assert(switchBinder is { });
            return switchBinder.BindSwitchExpressionCore(node, switchBinder, diagnostics);
        }
```

```
        private BoundExpression BindCheckedExpression(CheckedExpressionSyntax node, BindingDiagnosticBag diagnostics)
        {
            // the binder is not cached since we only cache statement level binders
            return this.WithCheckedOrUncheckedRegion(node.Kind() == SyntaxKind.CheckedExpression).
                BindParenthesizedExpression(node.Expression, diagnostics);
        }
```

### CODE REVIEW

**Comment by 333fred** [comment]
It would probably we worth checking other binder flags as well: is `unsafe` handled correctly, for example?

**Comment by jcouv** [comment]
> It would probably be worth checking other binder flags as well: is `unsafe` handled correctly, for example?

That's beyond the scope of this fix, but nothing jumps out from cursory glance at other flags (including `unsafe`, see `LocalBinderFactory.VisitUnsafeStatement`).

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Portable/Binder/LocalBinderFactory.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/LocalBinderFactory.cs:616]: I think the tests for #60419 are already checked in.

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:343]: >current is CheckedExpressionSyntax [](http://example.com/codeflow?start=25&length=34)

We probably should check that we are inside parens. We do similar check for argument list and switch section, for example. if we are on the checked/unchecked keyword, we haven't entered the context yet, that should be possible to observe through a test.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 4), assuming CI is passing

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

**Review by 333fred — APPROVED** [review]
LGTM (commit 4), assuming tests are passing.

**Pull request merged**
